op-dispute-mon/mon/resolution/delay.go

```diff
@@ -0,0 +1,53 @@
+package resolution
+
+import (
+	"github.com/ethereum-optimism/optimism/op-challenger/game/fault/types"
+	monTypes "github.com/ethereum-optimism/optimism/op-dispute-mon/mon/types"
+	"github.com/ethereum-optimism/optimism/op-service/clock"
+)
+
+type DelayMetrics interface {
+	RecordClaimResolutionDelayMax(delay float64)
+}
+
+type DelayCalculator struct {
+	metrics DelayMetrics
+	clock   clock.Clock
+}
+
+func NewDelayCalculator(metrics DelayMetrics, clock clock.Clock) *DelayCalculator {
+	return &DelayCalculator{
+		metrics: metrics,
+		clock:   clock,
+	}
+}
+
+func (d *DelayCalculator) RecordClaimResolutionDelayMax(games []monTypes.EnrichedGameData) {
+	var maxDelay uint64 = 0
+	for _, game := range games {
+		delay := d.getMaxDelay(&game)
+		if delay > maxDelay {
+			maxDelay = delay
+		}
+	}
+	d.metrics.RecordClaimResolutionDelayMax(float64(maxDelay))
+}
+
+func (d *DelayCalculator) getMaxDelay(game *monTypes.EnrichedGameData) uint64 {
+	var maxDelay uint64 = 0
+	for _, claim := range game.Claims {
+		delay := d.getDelay(&claim)
+		if delay > maxDelay {
+			maxDelay = delay
+		}
+	}
+	return maxDelay
+}
+
+func (d *DelayCalculator) getDelay(claim *types.Claim) uint64 {
+	// If the bond amount is the max uint128 value, the claim is resolved.
+	if monTypes.MaxUint128.Cmp(claim.ClaimData.Bond) == 0 {
+		return 0
+	}
+	return uint64(d.clock.Now().Unix() - int64(claim.Clock))
```

Comment:
This is the first time we've actually used the clock value and I don't think we're handling it correctly.  The contract clock is actually a combination of two values - the remaining duration and the timestamp of the last update: https://github.com/ethereum-optimism/optimism/blob/e4e02eb1c0741a692a0c226abf8a30414bd7d99b/packages/contracts-bedrock/src/libraries/DisputeTypes.sol#L59-L67

Whereas we're just treating it as a timestamp: https://github.com/ethereum-optimism/optimism/blob/215abc31277ade31e736d40dfc64f5ea64d38a7b/op-challenger/game/fault/contracts/faultdisputegame.go#L327

So we need to actually parse it as two separate values, then do something like:
```
timeSinceLastUpdate := d.clock.Now().Unix() - claim.ClockTimestamp
remainingDuration := claim.ClockDuration - timeSinceLastUpdate
```
And the resolution delay is `-1 * remainingDuration` (we can't resolve the claim until remainingDuration == 0).

I'd suggest adding a `Claim.RemainingDuration(now time.Time) time.Duration` function to calculate the remaining duration (might be easier for it to return `int64` seconds rather than `time.Duration` but would be nice if we can use the go types.

Reply:
So I discussed with @clabby a bit more. My understanding here is that we don't want to be subtracting the `timeSinceLastUpdate` from the `claim.ClockDuration` value. The `claim.ClockDuration` value is an accumulator over the chess time for the given claim. Thus the total amount of time used by the claim is `claim.ClockDuration + timeSinceLastUpdate`. Then, the time remaining is `(MAX_GAME_DURATION / 2) - (claim.ClockDuration + timeSinceLastUpdate)`. I believe in the calculation you performed, the `claim.ClockDuration` was incorrectly considered to represent how much time was remaining for the given chess clock at the time the claim was posted.

Correct me if I'm wrong here but I've updated the stack to compute the total chess time used by the claim given the current time (which is cleaner anyways since we avoid underflows). I've then added game duration fetching in this pr and it's use in the `DelayCalculator` to calculate how much time is remaining for a given claim **until** resolution (resolution delay ??).

So, this pr currently metrices the **maximum** amount of time until resolution for any given claim across all games. I'm thinking this is a misunderstanding and we want the metric to actually calculate the minimum amount of resolution time to track how long until the next claim will be resolved (?). I've stacked a pr ontop of this to add that metric. Since this is now as simple as taking a `min(a, b)` across all the remaining chess times of all claims in all games, it comes at zero cost as the `DelayCalculator` already computes the time remaining for each claim 😄 

Again, let me know if this is what you're thinking here.

Reply:
Opened the stacked, downstream pr here: https://github.com/ethereum-optimism/optimism/pull/9587

Reply:
hmm, so the metric we want is one that looks at each claim across all games and filters to exclude:
* claims that are not yet resolvable
* claims that have already been resolved

Then of the remaining claims report the maximum time since the claim could have been resolved. We're looking for cases where the challenger isn't doing its job and resolving claims quickly once they become resolvable.

We don't care about claims that haven't yet had their clocks expire.

Reply:
Ahh I understand what you're looking for now got it. I think your calculation for the "overflow time", taking `accumulatedTime - maxChessClockTime` is correct then for this metric.